src/artm/regularizer/net_plsa_phi.cc

```diff
@@ -0,0 +1,166 @@
+                               const ::artm::core::PhiMatrix& n_wt,
+                               ::artm::core::PhiMatrix* result) {
+  // read the parameters from config and control their correctness
+  const int topic_size = p_wt.topic_size();
+  const int token_size = p_wt.token_size();
+
+  std::vector<bool> topics_to_regularize;
+  if (config_.topic_name().size() == 0)
+    topics_to_regularize.assign(topic_size, true);
+  else
+    topics_to_regularize = core::is_member(p_wt.topic_name(), config_.topic_name());
+
+
+  if (!config_.has_class_id()) {
+    LOG(ERROR) << "There's no name of vertex modality in class_id field for" <<
+                  "NetPLSA regularizer. Cancel it's launch.";
+    return false;
+  }
+
+  bool has_weights = config_.vertex_weight_size();
+  if (has_weights && vertex_name2id_.size() != config_.vertex_weight_size()) {
+    LOG(ERROR) << "Non-empty vertex_weight array should have the same length " <<
+                  "with vertex_name array in NetPLSA regularizer config (" <<
+                  vertex_name2id_.size() << " != " << config_.vertex_weight_size() << ")";
+  }
+
+  std::vector<double> n_t(topic_size, 0.0);
+  for (int token_id = 0; token_id < token_size; ++token_id) {
+    const auto& token = n_wt.token(token_id);
+    if (token.class_id != config_.class_id())
+      continue;
+
+    for (int topic_id = 0; topic_id < topic_size; ++topic_id) {
+      if (topics_to_regularize[topic_id]) {
+        n_t[topic_id] += static_cast<double>(n_wt.get(token_id, topic_id));
+      }
+    }
+  }
+
+  for (int token_id = 0; token_id < token_size; ++token_id) {
+    const auto& token_u = p_wt.token(token_id);
+    if (token_u.class_id != config_.class_id())
+      continue;
+
+    auto name_iter = vertex_name2id_.find(token_u.keyword);
+    if (name_iter == vertex_name2id_.end())
+      continue;
+
+    auto edge_iter = edge_weights_.find(name_iter->second);
+    if (edge_iter == edge_weights_.end())
+      continue;
+
+    float D_u = has_weights ? config_.vertex_weight(name_iter->second) : 1.0;
+    for (int topic_id = 0; topic_id < topic_size; ++topic_id) {
+      if (!topics_to_regularize[topic_id])
+        continue;
+
+      float value = 0.0;
+      float p_ut = p_wt.get(token_id, topic_id);
+      for (auto iter = edge_iter->second.begin(); iter != edge_iter->second.end(); ++iter) {
+        auto it = id2vertex_name_.find(iter->first);
+        if (it == id2vertex_name_.end()) {
+          LOG(WARNING) << "Edge links to vertex " << iter->first <<
+                          ", that does not exist in list of vertices, it will be skipped";
+          continue;
+        }
+
+        float p_vt = p_wt.get(p_wt.token_index(
+            ::artm::core::Token(config_.class_id(), it->second)), topic_id);
+        float D_v = has_weights ? config_.vertex_weight(iter->first) : 1.0;
+
+        value += iter->second * (p_vt / (D_u * D_v) - p_ut / (D_u * D_u));
```

Comment:
Is it better to multiply by `1 / D_u` out of parentheses to avoid some problems with single precision?